### ISSUE TITLE
xpath_elements_to_hide not always working

### ISSUE THREAD

**Issue by boris-petrov** [jnormington/dotdiff#19]
I'm testing all those XPATHs in the browser with `$x("some-xpath")`. One example is `or` clauses:

```//div[contains(text(), 'a')]/following-sibling::div```
```//div[contains(text(), 'b')]/following-sibling::div```
```//div[contains(text(), 'c')]/following-sibling::div```

All of these work by themselves (both in the browser and in `xpath_elements_to_hide`). However if I do it the following way:

```//div[contains(text(), 'a') or contains(text(), 'b') or contains(text(), 'c')]/following-sibling::div```

Which is equivalent to the above three cases. It works in the browser but doesn't work when I put it in `xpath_elements_to_hide`, i.e. the elements are still visible in the screenshot.

Another example is this:

```//div[contains(@class, 'qwe')]/div[1]```

Not sure why it doesn't hide the element.

Any idea what I might be doing wrong or where the bug is?

**Comment by jnormington** [comment]
Hmm weird, I'm guessing there's no error ? 

I suspect its working but changing it to just one entry is causing some race condition, for example the DotDiff#max_wait_time is too low. We also in the past experienced this but I never looked into it as I opted for specific element snapshots ( you know this is also possible if that works better for you but then you can't use xpath_elements_to_hide)

Anyhow could you try this code below which is the steps it takes but we can inspect if the element is found then add some sleep to see where it is failing or not.

```ruby
xpath = "//div[contains(text(), 'a') or contains(text(), 'b') or contains(text(), 'c')]/following-sibling::div"
handler = DotDiff::ElementHandler.new(page, [xpath])

puts "ElementExists: #{!!handler.element_exists?}"

handler.hide
# sleep 2 # Try this after if the element exists but isn't hidden
page.save_screenshot("github_issue_test.png")
```

I haven't test the above code - but it should work

Try the above if the element isn't hidden in the image and if the ElementExists is true can you try adding a sleep statement before taking the screenshot as I have commented to see if adding some wait works.

Lets see what the outcome is and go from there.

**Comment by boris-petrov** [comment]
There is no error, no.

I tried your code (had to change `handler.element_exists?` to `handler.element_exists?(xpath)`) and it prints `false`. Which is funny because if I pause the test just after the `puts` and I do a `$x(same_xpath_as_in_the_rubu)` it returns me the three elements. :) So there is a different implementation of xpath in jQuery and in whatever you're using?

**Comment by jnormington** [comment]
We use capybara `find(:xpath, some_xpath)` to check if the element is present, it would need to deeper digging into how Capybara is handling the xpath if that is the issue (which from the output it isn't finding the element)

Can you also run the following after `handler.hide` this is using the native js to hide the element, but  bypasses the capybara check

```ruby
page.execute_script("#{handler.js_element(xpath)}.style.visibility = 'hidden'")
```

**Comment by boris-petrov** [comment]
This worked, in a way. It removed an element but only the first one in the page. I guess this is normal because I guess `js_element` returns a single element. So, again, probably the issue is in Capybara's `find`?

**Comment by jnormington** [comment]
@boris-petrov yes thats because the forth arg to document.evaluate only specify it will return the first element `XPathResult.FIRST_ORDERED_NODE_TYPE` - my initial intentions was for it to hide individual elements not many because it becomes a little more complex with waiting correctly.

So the issue is within Capybara but as dotdiff wasn't intended to support multiple elements in one call, and as it appears you might want multiple elements to be hidden I think the best way forward is to use your first approach.

There are definitely some improvements but there aren't currently any plans to work on it in the near future.

**Comment by boris-petrov** [comment]
Ok, thanks. I'm fine with using the first approach, however I have some other cases where an element is not found at all while it is by `$x(some_xpath)` and I'm not sure how to proceed there. Now that I think about it, I think these don't work:

```//div[contains(@class, 'a')]/div[1]```

I think the problem here is in `[1]`. I have another case with a `/span[2]` which also doesn't work so I guess that's the issue. I guess this, again, is in Capybara?

**Comment by jnormington** [comment]
@boris-petrov - the issue with elements not being found might be that the elements are not present immediately is this possible ? I mean what is your Dotdiff.max_wait_time config value. 

That xpath should be working just fine - I have match more complex xpath and others just like yours above.

One more thing to try is the following - because maybe some different error is occurring but rescue nil is catching everything (which should be improved on my part)

```ruby
xpath = "//div[contains(@class, 'a')]/div[1]"

handler = DotDiff::ElementHandler.new(page, [xpath])
elem = page.find(:xpath, xpath, wait: DotDiff.max_wait_time, visible: :all)

puts "Elem: #{!!elem}"
puts "ElementExists: #{!!handler.element_exists?}"

handler.hide
# sleep 2 # Try this after if the element exists but isn't hidden
page.save_screenshot("github_issue_test.png")
```
if it errors what is the error ? Based on what you have been saying I expect it will fail - likely the DotDiff.max_wait_time is too low (trying increasing).

**Comment by boris-petrov** [comment]
I tried your code. I changed `page.find` with `page.first` as otherwise it blows up with `Ambiguous match, found 3 elements matching xpath "//div[contains(@class, 'a')]/div[1]"`. After the change, the output is:

```
Elem: true
ElementExists: false
```

Which is not what one would expect I guess. :) Also, the elements are not hidden in the image. For the sake of the test, I put a `config.max_wait_time = 20`. Ideas? :)

**Comment by jnormington** [comment]
Hey @boris-petrov,

So now we know the issue - its that your xpath isn't targeting specific element and instead is returning multiple matches, this is the reason I asked for you to run the `page.find()` call. 

This is what the element handler is using and is expecting only to find one match which is the default config for Capybara.

Are you able to make to create xpath string to match just a single element ? I mean how can you be sure that it will always be the first ? That will become very brittle in the tests (assuming you are running this in test)

Anyhow if this is the required behavior, I believe the following Capybara config option is what you require

```ruby
Capybara.configure do |config|
  config.match = :first
end
```

If recall correctly that will change the match query and instead of erroring with Ambiguous error, it will instead return the first element by default just like your `page.first` call.

Once set the Capybara calls should use that config. Hope that helps, let me know how you get on. Think we are nearing the end.

**Comment by boris-petrov** [comment]
Thanks for the answer!

Am I understanding you right that the `xpath_elements_to_hide` option expects all xpaths to match just a **single** element? If so - why is that and how do I make it work for multiple? In my tests there are many cases when I want to hide many things with the same xpath! Am I misunderstanding something or is that a limitation of `dotdiff`?

**Comment by jnormington** [comment]
Hey @boris-petrov,

Yes this is a limitation of dotdiff, my use case was to only hide a specific elements which could be found with xpath. As mentioned earlier in the thread, this wasn't intended to support finding multiple elements with a single xpath query and hide them all. 

I think that you would likely be able to implement it but you would also need to pass how many elements you would be expecting to find as well, to ensure it waited for all to be present on the page (maybe you don't have this issue)

**Comment by jnormington** [comment]
Hey @boris-petrov 

I've had a look at this and I think I have created a working monkey patch, however it is untested so there might be some errors, anyhow the following should work for your use case.

It will use Capybara to create some wait time once it finds the first element it will return.
(I am assuming that in your use case once one element in the xpath query is found) all will be present on the page. If not you will need to work out how to pass down the count to Capybara)

It will return boolean and then it will execute js to get all elements matching the xpath query and iterate over them all and set the visibility.

You need to put the following content in an initializer so that it is loaded early on.

I hope that this helps. Again this code hasn't been ran so think this should work for you.

```ruby
require 'dotdiff'

module DotDiff
  class ElementHandler
    attr_accessor :driver

    def hide
      elements.each do |xpath|
        if elements_exists?(xpath)
          driver.execute_script(
            "var elems = #{js_element(xpath)}; "\
            "while(elem = elems.iterateNext()) { elem.style.visibility = 'hidden'; }"
          )
        end
      end
    end

    def show
      elements.each do |xpath|
        if elements_exists?(xpath)
          driver.execute_script(
            "var elems = #{js_element(xpath)}; "\
            "while(elem = elems.iterateNext()) { elem.style.visibility = ''; }"
          )
        end
      end
    end

    def js_elements(xpath)
      "document.evaluate(\"#{xpath}\", document, null, XPathResult.ORDERED_NODE_ITERATOR_TYPE, null)"
    end

    def elements_exists?(xpath)
      driver.all(:xpath, xpath, wait: DotDiff.max_wait_time, visible: :all).empty?
    end
  end
end
```

**Comment by boris-petrov** [comment]
Hi, thanks for the answers and the code! I'm a bit busy right now, I'll try it probably next week and notify you what's going on. Thanks again!

**Comment by boris-petrov** [comment]
@jnormington - why did you close this? I haven't forgotten it, it's just that I'm a bit overwhelmed at the moment and this is not a priority as I've worked it around. Are you planning on committing something to fix that?

**Comment by jnormington** [comment]
Hey @boris-petrov,

The reason I closed was there was nothing for me to do and the action remaining was for you to use the monkey patch I provided in above comment.

I don't think there was anything to fix - dotdiff was only ever intended to target a single element. To support multiple elements would be confusing and not straight forward...

You would need to specify how many elements we would be expecting for example if we have two elements with the same identifier and the 1st loads in under a second but the 2nd takes longer we would return as soon as one element is matching, therefore not hiding all the elements as one might expect. Thats because we need to know how many elements we would be expecting.

This would also make dotdiff quite brittle to any changes that might happen on the DOM for example if another element with the same identifier was added to the page that loaded in under a second it would return the 1st and 3rd element and the 2nd would causing some confusion.  

For that reason it wouldn't be something I would be looking to fix/implement and the monkey patch based on your scenario would/should appear to work for you just fine, although could also run into the issue I just mentioned.

Hope that helps

**Comment by boris-petrov** [comment]
Well, I always thought from a different perspective - dotdiff is not supposed to "wait" for anything or "check" anything - it's supposed to take a screenshot and compare it against some other. It's my job as a programmer to make sure *everything* I want to be visible to be there and *everything* I want hidden to also be there. Then when I call dotdiff, I would expect it to just hide everything from the list of selectors that is there and that's it. I have Capybara for the rest - to wait on elements to become visible for example. The race conditions you are talking about are a valid problem but again - it **should** not be dotdiff's responsibility to handle these but the programmer's. Do I make sense? :)

**Comment by jnormington** [comment]
I can see your point - but that wasn't my initial intention for full page screenshots, but this is how it works for single element snapshots, because they don't hide elements.

At this point its not something I intend to change or implement anytime soon, but feel free to file a  PR with your proposed additions/changes with specs.

The other thing is based on your previous message if you are already ensuring all the elements are  present you could just rip the code I created as a monkey patch and call it separately before calling dotdiff to do the snapshot, and create a small wrapper so in your specs you can still call it in a still a single line, and remove the list from the dotdiff config.

Then dotdiff is just a snapshot and compare library and nothing more for your use case.
